bpf/bpf_host.c

```diff
@@ -489,6 +489,37 @@ int tail_handle_ipv6_from_netdev(struct __ctx_buff *ctx)
 {
 	return tail_handle_ipv6(ctx, false);
 }
+
+# ifdef ENABLE_HOST_FIREWALL
+static __always_inline int
+handle_to_netdev_ipv6(struct __ctx_buff *ctx)
+{
+	void *data, *data_end;
+	struct ipv6hdr *ip6;
+	__u8 nexthdr, type;
+	__u32 srcID = 0;
+	int hdrlen;
+
+	if (!revalidate_data(ctx, &data, &data_end, &ip6))
+		return DROP_INVALID;
+
+	nexthdr = ip6->nexthdr;
+	hdrlen = ipv6_hdrlen(ctx, ETH_HLEN, &nexthdr);
+	if (hdrlen < 0)
+		return hdrlen;
+
+	if (unlikely(nexthdr == IPPROTO_ICMPV6)) {
+		type = icmp6_load_type(ctx, ETH_HLEN);
+		if (type == ICMP6_NS_MSG_TYPE || type == ICMP6_NA_MSG_TYPE)
+			return CTX_ACT_OK;
+	}
```

Comment:
Per my overall summary here:

https://github.com/cilium/cilium/pull/12049#issuecomment-643430893

I think we should actually be allowing also the router solicitation / router advertisements here, I don't think we can make many assumptions about the underlying network that would make it OK to be dropping such ICMPv6 peer discovery packets. Or even if we can, we should really think it through and design that as a feature.

Reply:
The summary makes sense. But how should we decide which ICMPv6 messages we want to let through besides the NS, NA, RS, RA? We probably don't want echo requests/replies because that would be inconsistent with how we handle ICMPv4, but what about more obscure (to me at least) cases like Router Renumbering or Inverse Neighbor Discovery?

I'm fine with letting through a minimal set of known required messages and then extending if we see the need for other cases. That's what I was going for, even though I had missed RSes and RAs.

Reply:
Turns out there's an RFC on this topic ;)

https://tools.ietf.org/html/rfc4890

Probably looking at Section 4.4 primarily though I didn't yet read through this document in detail.

This also states that echo request/response "must not be dropped", under the "recommendations" for ICMPv6 Local Configuration Traffic. I can tell you the argument for dropping such messages in the pods case which is basically that if you are not explicitly allowing L3+ traffic on a given interface, we logically disconnect the interface from the L3 network so it should be as though there is no connectivity to that interface. To be consistent with our pod network policy semantics I would assume that we take the same stance for host firewall.

I note also that for the case where we are doing host firewall and forwarding traffic to pods, it's likely that section 4.3, "Recommendations for ICMPv6 Transit Traffic" apply to us for that traffic.

Reply:
I spent some time reading the RFC this morning and I think we should process ICMPv6 message types as follows. Only our handling of `ICMPv6-echo-reply` and `ICMPv6-echo-message` would not be conforming to the RFC (when denied by host policies). All experimental, unallocated, and unassigned types would be rejected, as well as functions for mobility and `ICMPv6-redirect-message`. 

|          ICMPv6 Message         |   Action     | Type |         Comment         |
|---------------------------------|--------------|------|-------------------------|
|          ICMPv6-unreach         |  CTX_ACT_OK  |   1  |                         |
|          ICMPv6-too-big         |  CTX_ACT_OK  |   2  |                         |
|           ICMPv6-timed          |  CTX_ACT_OK  |   3  | Code 0 is MUST          |
|         ICMPv6-parameter        |  CTX_ACT_OK  |   4  | Code 1 and 2 are MUST   |
|    ICMPv6-err-private-exp-100   | CTX_ACT_DROP |  100 | Experimental            |
|    ICMPv6-err-private-exp-101   | CTX_ACT_DROP |  101 | Experimental            |
|       ICMPv6-err-expansion      | CTX_ACT_DROP |  127 |                         |
|       ICMPv6-echo-message       | Sent to firewall |  128 | Despite recommandation  |
|        ICMPv6-echo-reply        | Sent to firewall |  129 | Despite recommandation  |
|      ICMPv6-mult-list-query     |  CTX_ACT_OK  |  130 |                         |
|      ICMPv6-mult-list-report    |  CTX_ACT_OK  |  131 |                         |
|      ICMPv6-mult-list-done      |  CTX_ACT_OK  |  132 |                         |
|    ICMPv6-mult-list-report-v2   |  CTX_ACT_OK  |  143 |                         |
|      ICMPv6-router-solici       |  CTX_ACT_OK  |  133 |                         |
|      ICMPv6-router-advert       |  CTX_ACT_OK  |  134 |                         |
|     ICMPv6-neighbor-solicit     |  CTX_ACT_OK  |  135 |                         |
|      ICMPv6-neighbor-advert     |  CTX_ACT_OK  |  136 |                         |
|     ICMPv6-redirect-message     | CTX_ACT_DROP |  137 | Potential security risk |
|      ICMPv6-router-renumber     |  CTX_ACT_OK  |  138 | Uses IPSec anyway       |
|      ICMPv6-node-info-query     | CTX_ACT_DROP |  139 | Experimental            |
|     ICMPv6-node-info-response   | CTX_ACT_DROP |  140 | Experimental            |
|   ICMPv6-inv-neighbor-solicit   |  CTX_ACT_OK  |  141 |                         |
|    ICMPv6-inv-neighbor-advert   |  CTX_ACT_OK  |  142 |                         |
| ICMPv6-home-agent-disco-request | CTX_ACT_DROP |  144 | Needed for mobility     |
|  ICMPv6-home-agent-disco-reply  | CTX_ACT_DROP |  145 | Needed for mobility     |
|      ICMPv6-mobile-solicit      | CTX_ACT_DROP |  146 | Needed for mobility     |
|      ICMPv6-mobile-advert       | CTX_ACT_DROP |  147 | Needed for mobility     |
|      ICMPv6-send-solicit        |  CTX_ACT_OK  |  148 |                         |
|       ICMPv6-send-advert        |  CTX_ACT_OK  |  149 |                         |
|       ICMPv6-mobile-exp         | CTX_ACT_DROP |  150 | Experimental            |
|    ICMPv6-mult-router-advert    |  CTX_ACT_OK  |  151 |                         |
|    ICMPv6-mult-router-solicit   |  CTX_ACT_OK  |  152 |                         |
|     ICMPv6-mult-router-term     |  CTX_ACT_OK  |  153 |                         |
|         ICMPv6-FMIPv6           | CTX_ACT_DROP |  154 |                         |
|       ICMPv6-rpl-control        | CTX_ACT_DROP |  155 |                         |
|   ICMPv6-info-private-exp-200   | CTX_ACT_DROP |  200 | Experimental            |
|   ICMPv6-info-private-exp-201   | CTX_ACT_DROP |  201 | Experimental            |
|      ICMPv6-info-expansion      | CTX_ACT_DROP |  255 | Future use              |
|       ICMPv6-unallocated        | CTX_ACT_DROP |      | Unallocated             |
|       ICMPv6-unassigned         | CTX_ACT_DROP |      | Unassigned types        |

Reply:
FWIW I think the intent of the existing BPF policy code is that if there's something like a dst unreachable, we will relate that back to the original outbound flow and allow it as a "related response", implying that we don't have to explicitly allow all such responses. This is achieved through inserting a CT entry matching ICMP in the reverse direction whenever we insert a CT entry on the outbound direction for a TCP/UDP flow.